doop.c

```diff
@@ -657,17 +688,72 @@ Magic and tainting are handled.
 =cut
 */
 
+static SV *
+S_concat_maybe_overloaded(pTHX_ SV *lhs, SV *rhs)
+{
+    STRLEN len;
+    const char *s;
+
+    /* We need both ROK and RMG in order to even have a chance of having
+     * concat overload, so we might be able to speed this up with: */
+    /* U32 flags = SvFLAGS(lhs)|SvFLAGS(rhs); */
+    /* if (UNLIKELY((flags & (SVf_ROK|SVs_GMG)))) { */
+    if (UNLIKELY(S_sv_has_magic_method(aTHX_ lhs, concat_amg)
+              || S_sv_has_magic_method(aTHX_ rhs, concat_amg))) {
+        dSP;
+        EXTEND(SP, 2);
+        PUSHs(lhs);
+        PUSHs(rhs);
+        PUTBACK;
+
+        if (Perl_try_amagic_bin(aTHX_ concat_amg, 0)) {
+            SPAGAIN;
+            SV *ret = POPs;
+            PUTBACK;
+
+            return ret;
+        }
+
+        SPAGAIN;
+        PERL_UNUSED_RESULT(POPs);
+        PERL_UNUSED_RESULT(POPs);
+        PUTBACK;
+    }
+    s = SvPV_const(rhs, len);
+    sv_catpvn_flags(lhs, s, len, DO_UTF8(rhs) ? SV_CATUTF8 : SV_CATBYTES);
+
+    return lhs;
+}
+
 void
 Perl_do_join(pTHX_ SV *sv, SV *delim, SV **mark, SV **sp)
 {
     SV ** const oldmark = mark;
+    SV *orig_sv = sv;
+    SV *delimsv = NULL;
+    SV *delimcopy = NULL;
+    int delim_has_concat = S_sv_has_magic_method(aTHX_ delim, concat_amg);
```

Comment:
This checks for AMAGIC before fetching normal magic, giving inconsistent results (see below).

I think it has similar problems for magic values in the LIST.

```
tony@venus:.../git/perl6$ ./perl -Ilib ../20503b.pl
A.helloB.helloC.helloD
AwrongBwrongCwrongD
A.helloB.helloC.helloD
tony@venus:.../git/perl6$ cat ../20503b.pl
use v5.36.0;

my $bobj = tie my $x, "Foo";

my $y = Bar->new("hello");
$$bobj = $y;

my @l = "A" .. "D";

say join $y, @l;

say join $x, @l;
say join $x, @l;

package Foo;

sub TIESCALAR {
    my ($c, $v) = @_;
    bless \$v, $c;
}

sub FETCH {
    ${$_[0]};
}

package Bar;

sub new {
    my ($c, $t) = @_;

    bless \$t, $c;
}

use overload '.' =>
  sub {
      my ($l, $r, $sw) = @_;

      my $lt = ref $l eq "Bar" ? $$l : "$l";
      my $rt = ref $r eq "Bar" ? $$r : "$r";

      return $sw ? "$rt.$lt" : "$lt.$rt";
  },
  '""' => sub { "wrong" };
```

Reply:
Thank you, this issue is now addressed in the latest version of the branch.